sdk/core/Azure.Core.Experimental/src/Primitives/BinaryData.cs

```diff
@@ -53,64 +56,120 @@ public BinaryData(ReadOnlySpan<byte> data)
         /// </summary>
         /// <param name="data">Byte data.</param>
-        private BinaryData(ReadOnlyMemory<byte> data)
+        public BinaryData(byte[] data)
         {
             Bytes = data;
         }
 
         /// <summary>
-        /// Creates a binary data instance from a string by converting
-        /// the string to bytes using UTF-8 encoding.
+        /// Creates a <see cref="BinaryData"/> instance by serializing the passed in object to Json
+        /// using <see cref="JsonObjectSerializer"/>.
         /// </summary>
-        /// <param name="data">The string data.</param>
+        ///
+        /// <param name="jsonSerializable">The object that will be serialized to Json using
+        /// <see cref="JsonObjectSerializer"/>.</param>
+        /// <param name="type">The type of the data. If not specified, <see cref="object.GetType"/> will
+        /// be used to determine the type.</param>
+        ///
         /// <returns>A <see cref="BinaryData"/> instance.</returns>
-        /// <remarks>The byte order mark is not included as part of the encoding process.</remarks>
-        public BinaryData(string data)
+        public BinaryData(object jsonSerializable, Type? type = default)
+            : this(jsonSerializable, s_jsonSerializer, type)
         {
-            Bytes = s_encoding.GetBytes(data);
         }
 
         /// <summary>
-        /// Creates a binary data instance from an object and serializes it
+        /// Creates a <see cref="BinaryData"/> instance by serializing the passed in object
         /// using the provided <see cref="ObjectSerializer"/>. If no <see cref="ObjectSerializer"/>
         /// is specified, <see cref="JsonObjectSerializer"/> will be used.
         /// </summary>
-        /// <param name="data">The data that will be serialized.</param>
-        /// <param name="type">The type of the data.</param>
-        /// <param name="serializer">The serializer to serialize
-        /// the data.</param>
+        ///
+        /// <param name="serializable">The data that will be serialized.</param>
+        /// <param name="serializer">The serializer to serialize the data.</param>
+        /// <param name="type">The type of the data. If not specified, <see cref="object.GetType"/> will
+        /// be used to determine the type.</param>
+        ///
         /// <returns>A <see cref="BinaryData"/> instance.</returns>
-        public BinaryData(object data, Type type, ObjectSerializer? serializer = default)
+        public BinaryData(object serializable, ObjectSerializer serializer, Type? type = default)
         {
             using var memoryStream = new MemoryStream();
-            serializer ??= new JsonObjectSerializer();
-            serializer.Serialize(memoryStream, data, type, CancellationToken.None);
+            serializer ??= s_jsonSerializer;
+            serializer.Serialize(memoryStream, serializable, type ?? serializable.GetType(), CancellationToken.None);
```

Comment:
What if `serializable` is null?  Should it be `type ?? serializable?.GetType() ?? typeof(object)`?

Reply:
yikes! I think we can just not allow null objects.